Command line tools

subdavis opened the pull request:
* Adds coco and viame to dive conversion
* Adds dataset generation
* Adds DIVE json file verification against pydantic schema.

BryonLewis commented:
I don't know if you want this updated version of `generateLargeDatasey.py`, No guarantees on functionality but it allows for creation of videos as well, and I now modified the image/video so they have some noise to display.  I used this to generate data for the long loading PR testing.

https://gist.github.com/BryonLewis/ac74e93f954eba591ed76cd84c954be4

Updates:
- `--video` option to indicate you want a video
- `--width/--height` options to specify size of image or  size of video, all bboxes are still full frame though
- `--fps` indicator for video framerate.

subdavis commented:
This is probably going to go stale for a while, I'll deal with updating it later.

f4str commented on server/scripts/cli.py:
```diff
@@ -0,0 +1,86 @@
+    click.echo(f'wrote attrib {output_attrs.name}')
+
+
+@convert.command(name="viame2dive")
```
It might also be nice to also add a `dive2viame` command since the export csv serializer already exists.

f4str commented on server/scripts/generateLargeDataset.py:
```diff
@@ -1,24 +1,21 @@
 import cv2
 import numpy as np
-from PIL import Image
 from tqdm import tqdm
```
`numpy` and `tqdm` should probably also be added to the dev requirements.

f4str commented on server/scripts/cli.py:
```diff
@@ -0,0 +1,86 @@
+
+from dive_server.serializers import kwcoco, viame
+from dive_utils import models
+from scripts import generateLargeDataset
```
`generateLargeDataset` imports `numpy`, `opencv`, and `tqdm` and since `cli.py` imports `generateLargeDataset`, it also require those imports. This isn't an issue if the dev requirements are installed, but if the user wants to use the CLI, they would need the dev requirements.

f4str commented on server/setup.py:
```diff
@@ -58,6 +60,7 @@
     version="1.4.1",
     zip_safe=False,
     entry_points={
+        "console_scripts": ["dive=scripts.cli:cli"],
```
We should probably update the README with a really brief section that the CLI exists and how to use it.

BryonLewis commented on server/dive_utils/models.py:
```diff
@@ -43,6 +44,23 @@ class Track(BaseModel):
 
+    @validator('features')
+    def validateFeatures(cls, v: List[Feature], values: dict):
+        if len(v) > 0:
```
do we really allow a track that has no features?  It will pass validation but won't load.

subdavis replied:
Yes, because there are several places in code where the track object is initialized before features are added, such as in the deserializers.  see `load_csv_as_tracks_and_attributes`.

BryonLewis replied:
:facepalm: - was thinking primarily client side forgot about the creation of empty tracks during loading.

subdavis commented:
Updated changes:

* move serializers to `dive_utils` to prevent the command line from depending on the `dive_server` package (and because I think this is the correct place to put it anyway)
* Add dev and normal command lines, so that we can build a standalone that doesn't require cv2 and numpy

f4str commented on server/setup.py:
```diff
@@ -47,6 +51,7 @@
     include_package_data=True,
     zip_safe=False,
     entry_points={
+        "console_scripts": ["dive=scripts.cli_dev:cli"],
```
This is using the dev cli. We should probably have two entry points for console scripts (one for standard and one for dev) like this:

```python
"console_scripts": ["dive=scripts.cli:cli", "dive-dev=scripts.cli_dev:cli"],
```

subdavis replied:
There's no reason to expose both, IMO.  The dev cli includes everything.  The limited CLI is just so pyinstaller doesn't bundle numpy

f4str replied:
Got it. In that case, we should probably specify somewhere that the cli requires dev requirements.

f4str commented on server/tox.ini:
```diff
@@ -54,6 +54,14 @@ deps =
 
+[testenv:buildcli]
+extra =
+    dev
```
If we're packing the non-dev cli, there's no reason to include the dev requirements. This also brings the question if you should be able to build just one or both. `tox` can probably have a switch flag to choose. Alternatively, we can have two different tox options (`buildcli` and `buildcli_dev`).

subdavis replied:
Should only be able to build the non-dev

f4str replied:
Got it. Then you can just remove the 

```ini
extra =
     dev
```

This won't make a difference in the pyinstaller output, but by doing this `tox` won't need to install numpy and opencv just to setup the environment (which isn't needed for creating the non-dev cli). No effect on functionality, but will save some time.

f4str commented on server/tox.ini:
```diff
@@ -54,6 +54,14 @@ deps =
+deps =
+    pyinstaller
+commands =
+    pyinstaller --onefile scripts/cli.py
```
This doesn't seem to work correctly since the generated output file doesn't have any options other than `--help` and `--version`. I'm not sure if this still WIP or actually has a bug. I'm suspecting it might be something to do with the `cli.py` file not importing stuff correctly for pyinstaller to use in click.

Also be sure to add `{posargs}` at the end so additional optional flags can be passed.

subdavis replied:
Had to blow away my tox env to reproduce.  `rm -rf .tox && tox -e buildcli && ./dist/cli` .  Something I had before was making this work.

f4str commented on server/scripts/cli.py:
```diff
@@ -0,0 +1,15 @@
+    pass
+
+
+import scripts.commands  # noqa: E402 F401
```
This is kind of hacky to import all the click commands and rely on the import registering all the click commands. There are much better ways to separate out the commands and cli files. 

The top answer in this stackoverflow post is pretty good:
<https://stackoverflow.com/questions/34643620/how-can-i-split-my-click-commands-each-with-a-set-of-sub-commands-into-multipl>

But I'd recommend doing something similar to what girder is doing:
<https://github.com/girder/girder/tree/master/girder/cli>

This also might be the reason why the pyinstaller build doesn't work.

subdavis commented on server/scripts/generateLargeDataset.py:
```diff
@@ -0,0 +1,113 @@
+            tracks[track] = track_obj
+    json_file = f"{directory}/result_test.json"
+    with open(json_file, "w") as outfile:
+        json.dump(tracks, outfile)
```
should probably remove from both, it adds a lot of bytes to the file

f4str replied:
Agreed, it's probably better to have all json outputs be 1 line and compressed so people's text editors don't crash when opening massive files.

f4str commented:
This is great. I was emphasizing the `{posargs}` in tox so you can specify additional parameters. For example
```bash
tox -e buildcli -- --distpath path/to/output --clean --log-level DEBUG
```
Everything else seems to work great. I ran some tests and validation, conversion, and data generation all work perfectly. 

Just a few minor points
* As mentioned before, we should decide whether to remove the `indent=4` from the `viame2dive` and `coco2dive` so the files don't get bloated
* I noticed running 
  ```bash
  dive --version
   ```
  doesn't seem to work and crashes. Looks like click can't figure out the package distribution correctly for the `server/scripts` directory or something. Probably doesn't matter too much, but the work around is really simple. In `server/scripts/__init__.py` make the replacement
  ```python
  @click.version_option(1.0)
  ```
  to manually pass in a version number.
* `sever/testutils` is pretty much empty except for 2 json files which are only used in unit tests. One option is to just move them to `server/tests`, but this isn't that big of an issue.

f4str commented:
Update on the `--version` flag issue. Much simpler fix. According to <https://click.palletsprojects.com/en/8.0.x/api/#click.version_option> just specify the python package name

```python
@click.group()
@click.version_option(package_name='dive_server')
def cli():
    pass
```

subdavis commented:
@f4str I addressed those comments.

* `indent=4` was kept for the attributes files and removed for the track files.  The attributes files will always be small.
* testutils should just be json files now

f4str reviewed (approved):
Looks great! Tested it out (including `--version`) and everything runs without issue).

subdavis commented:
@BryonLewis would you like to review again?
